DDF-2820 Query feedback form for Intrigue

troymohl opened the pull request:
#### What does this PR do?
Adds query feedback form to Intrigue. Feedback is sent via Email to the configured email recipient. 

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@bdeining 
@rzwiefel 
@glenhein 

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[UI](https://github.com/orgs/codice/teams/ui)

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@andrewkfiedler
@millerw8

#### How should this be tested? (List steps with links to updated documentation)
1. Configure the email service to point to SMTP server
2. Configure Catalog UI Search, enabling Query Feedback and choosing a destination email that you wish to receive the feedback data at.
3. Perform a search on any workspace and click the 3 dots on the results tab
4. Choose `Submit Feedback` and submit feedback on the query
5. Check the email configured in Step 2 and verify that query feedback has been received.

#### Any background context you want to provide?

#### What are the relevant tickets?
[](https://codice.atlassian.net/browse/)
#### Screenshots (if appropriate)
Admin UI:
<img width="592" alt="screen shot 2017-04-05 at 1 40 59 pm" src="https://cloud.githubusercontent.com/assets/17255437/24723656/96ec4c14-1a05-11e7-89c2-384150290856.png">

Results menu option:
<img width="200" alt="screen shot 2017-04-05 at 1 40 07 pm" src="https://cloud.githubusercontent.com/assets/17255437/24723665/a19bbf1e-1a05-11e7-8785-e841627fcb10.png">

Feedback form:
<img width="1396" alt="screen shot 2017-04-05 at 1 40 37 pm" src="https://cloud.githubusercontent.com/assets/17255437/24723673/a6ec0352-1a05-11e7-9a0a-fe990cddbcc8.png">

Example email content:
<img width="701" alt="screen shot 2017-04-05 at 1 42 44 pm" src="https://cloud.githubusercontent.com/assets/17255437/24723686/b7909eac-1a05-11e7-875e-3d6e16e80652.png">

#### Checklist:
- [ ] Documentation Updated
- [X] Change Log Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/FeedbackApplication.java:
```diff
@@ -0,0 +1,264 @@
+
+    private static final ObjectMapper MAPPER = JsonFactory.create();
+
+    private ConfigurationApplication configurationApplication = null;
```
nit : do we need to explicitly set these to null?

troymohl replied:
Removed the explicit initialization.

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/FeedbackApplication.java:
```diff
@@ -0,0 +1,264 @@
+                res.body("{}");
+                res.status(200);
+                return res;
+            } catch (Exception e) {
```
Any chance at catching specific exceptions here?

troymohl replied:
Removed this completely as nothing is throwing a checked exception in this code.

bdeining commented on catalog/ui/catalog-ui-search/pom.xml:
```diff
@@ -327,6 +350,7 @@
                             !org.apache.commons.cli.*,
                             !org.hibernate.*,
                             !weblogic.jdbc.*,
+                            !org.abego.treelayout.*,
```
It looks like `org.abego.treelayout.*` is excluded in the dependencies, does this need to be omitted here too?

troymohl replied:
Still needed or the wiring fails in Karaf due to trying to find `org.abego.treelayout`

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/FeedbackApplication.java:
```diff
@@ -0,0 +1,264 @@
+                Session emailSession = smtpClient.createSession();
+                MimeMessage message = new MimeMessage(emailSession);
+                message.addRecipient(Message.RecipientType.TO, new InternetAddress(emailDestination));
+                message.setSubject(emailSubject);
```
If getEmailSubject can be null, is there a chance for an NPE on this line?

troymohl replied:
Per http://docs.oracle.com/javaee/6/api/javax/mail/internet/MimeMessage.html#setSubject(java.lang.String) I think null would be ok, as it would just not set the subject.

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/FeedbackApplication.java:
```diff
@@ -0,0 +1,264 @@
+
+                String emailSubject = getEmailSubject(feedback);
+                String emailBody = getEmailBody(feedback);
+                emailBody = emailBody.replaceAll("\\\\n", "\n");
```
If emailBody can be null, is there a chance for an NPE on this line?

troymohl replied:
Good catch. I'll add some null checks there in case we get null back from metatype/blueprint.

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/FeedbackApplication.java:
```diff
@@ -0,0 +1,264 @@
+
+import spark.servlet.SparkApplication;
+
+public class FeedbackApplication implements SparkApplication {
```
Is it possible to add unit tests to this class?

glenhein commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/FeedbackApplication.java:
```diff
@@ -0,0 +1,264 @@
+        initTemplates(true);
+    }
+
+    private static FeedbackRequest parseFeedbackRequest(String json) {
```
There are a lot of casts in this method. Are you sure they are all safe and guaranteed to work?

troymohl replied:
Updated to check before blindly casting.

glenhein commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/FeedbackApplication.java:
```diff
@@ -0,0 +1,264 @@
+        return subject;
+    }
+
+    private String getEmailBody(FeedbackRequest feedback) {
```
This method is nearly identical to getEmailSubject. Can you refactor them to remove the common code?

troymohl replied:
Refactored these methods.

glenhein commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/feedback/FeedbackRequest.java:
```diff
@@ -0,0 +1,120 @@
+ */
+package org.codice.ddf.catalog.ui.query.feedback;
+
+public class FeedbackRequest {
```
Could you add some javadoc to the class?

glenhein replied:
Some of the methods in this class are self-explanatory, but some are not, like getQuery. What does the return value for getQuery mean? Some javadoc on those methods would be nice.

troymohl replied:
Added javadocs

glenhein commented on catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.config.xml:
```diff
@@ -171,6 +171,35 @@
             default="25"
             required="true"/>
+
+        <AD id="queryFeedbackEnabled"
```
You might want to talk to @ricklarsen  about updating/improving the documentation for these new metatype fields.

troymohl replied:
Updated the config UI adoc file.

glenhein commented on catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.js:
```diff
@@ -91,6 +91,9 @@ define([
                     this.set('calculatedType', 'location');
                     break;
+                case 'TEXTAREA':
+                    this.set('calculatedType', 'textarea');
```
'calculatedType' is repeated several times. Should it be made into a constant?

andrewkfiedler replied:
I'd prefer it go like this if we want to refactor it:
```
var calculatedType;
switch(this.get('type')){
   case 'DATE':
          calculatedType = 'date';
          break;
   ... and so on
}
this.set('calculatedType', calculatedType);
```

bdeining commented on catalog/ui/catalog-ui-search/src/main/webapp/component/input/textarea/input-textarea.hbs:
```diff
@@ -0,0 +1,21 @@
+ **/
+ --}}
+<div class="if-editing">
+    <textarea id="{{cid}}" value="{{value}}" placeholder="{{property.placeholder}}" name="{{property.id}}"></textarea>
```
Is there any risk of code injection into this textarea?

andrewkfiedler replied:
Looks fine to me (at least on the side).  It's possible their mail service could be susceptible to an xss attack though, so it might not hurt to sanitize it for their sake.

andrewkfiedler replied:
Looks fine to me (at least on our side). It's possible their mail service could be susceptible to an xss attack though, so it might not hurt to sanitize it for their sake.

glenhein replied:
Since the email is going to show up in someone's inbox, most likely a web-based inbox, then xss could be a concern.

troymohl replied:
Fixed here: https://github.com/codice/ddf/pull/1858/files#diff-fb3aac9355e8ae0774fc8c12bba5c846R280

glenhein commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/feedback/FeedbackRequest.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    /**
+     * Set the workspace id that the feedback comments are being provided from.
```
You ended that sentence with a preposition! ;-) j/k

troymohl replied:
Fixed

troymohl commented:
build now

millerw8 commented:
build now

glenhein commented:
While manually testing this, I added comments that contained html sequences. They were escaped as expected, but I was hoping that the received email was also html so that my comments were displayed as I typed them, even though they are safely escaped.
![screen shot 2017-04-12 at 7 57 34 am](https://cloud.githubusercontent.com/assets/17837152/24964373/f99f379a-1f55-11e7-99f6-1fcd343ac7d4.png)

troymohl commented:
build now

andrewkfiedler commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/FeedbackApplication.java:
```diff
@@ -0,0 +1,289 @@
+        valueMap.put("username", feedbackRequest.getUsername());
+        valueMap.put("email", feedbackRequest.getEmail());
+        valueMap.put("workspace_id", feedbackRequest.getWorkspaceId());
+        valueMap.put("workspace_name", feedbackRequest.getWorkspaceName());
```
There's more here that needs the escapeHTML4.  Workspace name, query names, even the results themselves.

troymohl replied:
Shouldn't those have been escaped elsewhere in the UI code?

troymohl replied:
Fixed

troymohl commented:
@glenhein I have updated the email to send as HTML and fixed the template to default to simple HTML. Attached is what the current output looks like.
<img width="1383" alt="screen shot 2017-04-12 at 3 12 38 pm" src="https://cloud.githubusercontent.com/assets/17255437/24979761/abe04e12-1f92-11e7-9999-f3de330dc07b.png">

troymohl commented:
build now

andrewkfiedler commented on catalog/ui/catalog-ui-search/src/main/webapp/component/query-feedback/query-feedback.view.js:
```diff
@@ -0,0 +1,96 @@
+                region.currentView.turnOnEditing();
+            }
+        });
+        var tabable =  _.filter(this.$el.find('[tabindex], input, button, textarea'), function(element){
```
You can remove this section involving tabable.

troymohl replied:
Fixed.
